Cesium3DTilesContent/src/I3dmToGltfConverter.cpp

```diff
@@ -0,0 +1,738 @@
+  quaternion, in a way analagous to finding the half vector between two 3D
+  vectors.
+ */
+
+glm::quat rotation(const glm::vec3& vec1, const glm::vec3& vec2) {
+  float cosRot = dot(vec1, vec2);
+  // Not using epsilon for these tests. If abs(cosRot) < 1.0, we can still
+  // create a sensible rotation.
+  if (cosRot >= 1.0f) {
+    // zero rotation
+    return glm::quat(1.0, 0.0f, 0.0f, 0.0f);
+  }
+  if (cosRot <= -1.0f) {
+    // Choose principal axis that is "most orthogonal" to the first vector.
+    glm::vec3 orthoVec(0.0f, 0.0f, 0.0f);
+    glm::vec3 absVals(std::fabs(vec1.x), std::fabs(vec1.y), std::fabs(vec1.z));
+    if (absVals.z <= absVals.x && absVals.z <= absVals.y) {
+      orthoVec.z = 1.0f;
+    } else if (absVals.x <= absVals.y) {
+      orthoVec.x = 1.0f;
+    } else {
+      orthoVec.y = 1.0f;
+    }
+    auto rotAxis = cross(vec1, orthoVec);
+    rotAxis = normalize(rotAxis);
+    // rotation by pi radians
+    return glm::quat(0.0f, rotAxis.x, rotAxis.y, rotAxis.z);
+  }
+  auto rotAxis = cross(vec1, vec2);
+  glm::quat sumQuat(cosRot + 1.0f, rotAxis.x, rotAxis.y, rotAxis.z);
+  return normalize(sumQuat);
+}
+
+glm::quat rotationFromUpRight(const glm::vec3& up, const glm::vec3& right) {
+  // First rotation: up
+  auto upRot = rotation(glm::vec3(0.0f, 1.0f, 0.0f), up);
+  // We can rotate a point vector by a quaternion using q * (0, v) *
+  // conj(q). But here we are doing an inverse rotation of the right vector into
+  // the "up frame."
+  glm::quat temp =
+      conjugate(upRot) * glm::quat(0.0f, right.x, right.y, right.z) * upRot;
+  glm::vec3 innerRight(temp.x, temp.y, temp.z);
+  glm::quat rightRot = rotation(glm::vec3(1.0f, 0.0f, 0.0f), innerRight);
+  return upRot * rightRot;
+}
+
+struct ConvertResult {
+  GltfConverterResult gltfResult;
+  DecodedInstances decodedInstances;
+};
+
+/* The approach:
+  + Parse the i3dm header, decoding and creating all the instance transforms.
+  This includes "exotic" things like OCT encoding of rotations and ENU rotations
+  for each instance.
+  + For each node with a mesh (a "mesh node"), the instance transforms must be
+    transformed into the local coordinates of the mesh and then stored in the
+    EXT_mesh_gpu_instancing extension for that node. It would be nice to avoid a
+    lot of duplicate data for mesh nodes with the same chain of transforms to
+    the tile root. One way to do this would be to store the accessors in a hash
+    table, hashed by mesh transform.
+  + Add the instance transforms to the glTF buffers, buffer views, and
+    accessors.
+  + Metadata / feature id?
+*/
+
+CesiumAsync::Future<ConvertResult> convertInstancesContent(
+    const gsl::span<const std::byte>& instancesBinary,
+    const InstancesHeader& header,
+    uint32_t headerLength,
+    const CesiumGltfReader::GltfReaderOptions& options,
+    const ConverterSubprocessor& subprocessor,
+    GltfConverterResult& result) {
+  ConvertResult subResult;
```

Comment:
If you rename `ConvertResult` -> `ConvertedI3dm`, then could you also rename this?

```suggestion
  ConvertedI3dm convertedI3dm;
```

It's more intuitive in my head to see `convertedI3dm.gltfResult` than `subResult.gltfResult`